### PR TITLE
Add methods and tests to the AppUser model

### LINKED ISSUE #69: Add functions and tests for the AppUser model

### PR DESCRIPTION
**Pull request by LeonRabinovich**
### See the previous conversation of this PR: #71 
I lost the previous branch for this PR, and i had problems pushing to that PR from a different branch.
Hence - I decided to close the previous PR and open a new one.
____
Added the following AppUser functions:

- **toggle_user_activation**: toggles the "is_active" attribute of the user
- **create_appUser**: create an appUser and save it in the DB
- **get_all_appUsers**: returns a list of all AppUsers
- **find_appUser**: find an AppUser by username

Resolves: #69

### CODE REVIEW

File: homepage/tests/appUser_model_tests.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        assert appUsers[0].user.is_active is False
+
+        appUsers[0].toggle_user_activation()
+        assert appUsers[0].user.is_active is True
+
+    def test_get_all_appUsers(self, appUsers):
+        all_AppUsers = AppUser.get_all_appUsers()
+        assert (appUsers[0] in all_AppUsers
+                and appUsers[1] in all_AppUsers
+                and appUsers[2] in all_AppUsers)
```
Comment (oriAdler) [inline homepage/tests/appUser_model_tests.py]:
```suggestion
for appUser in appUsers:
    assert appUser in all_AppUsers
```

File: homepage/tests/appUser_model_tests.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        appUser1 = AppUser.create_appUser('user1', 'user1@mta.ac.il', 'useruser222')
+        appUser2 = AppUser.create_appUser('user2', 'user2@mta.ac.il', '1234')
+        appUser3 = AppUser.create_appUser('user3', 'user3@gmail.com', '10101010')
+
+        return (appUser1, appUser2, appUser3)
+
+    # This method tests:
+    # 1. The appUser created is the correct type of object (AppUser)
+    # 2. The appUser was created sucesfully and saved in the DB
+    def test_create_appUser(self, appUsers):
```
Comment (oriAdler) [inline homepage/tests/appUser_model_tests.py]: Maybe you should check a failure scenario when one of the arguments passed to ```AppUser.create_appUser``` isn't valid,
and then catch the exception from ```appUser.save()```

File: homepage/tests/appUser_model_tests.py (hunk 1 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: Changed

File: homepage/tests/appUser_model_tests.py (hunk 2 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: Added

**Review by oriAdler — APPROVED** [review]

**Review by ellaml — APPROVED** [review]

File: homepage/models.py [hunk 3]
```diff
@@ -109,11 +110,39 @@ def does_course_have_prerequisites(course):
...
+    @staticmethod
+    # Creates a new appUser (without superuser permmisions) and saves it in the DB
+    # The is_active attribute of AppUser.user is automaticly set to 'True' when creating a new AppUser
+    def create_appUser(username, email, password):
+        appUser = AppUser()
+        user = User.objects.create_user(username, email, password)
+        appUser.user = user
+
+        appUser.save()
+        return appUser
```
Comment (kobihk) [inline homepage/models.py]: maybe you can create new AppUser object by:
```suggestion
def __init__(self, username, email, password):
    self.user = User.objects.create_user(username, email, password)
    self.save()
```

File: homepage/models.py [hunk 4]
```diff
@@ -109,11 +110,39 @@ def does_course_have_prerequisites(course):
...
+
+    @staticmethod
+    def get_all_appUsers():
```
Comment (kobihk) [inline homepage/models.py]: python convention is:
https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables
```suggestion
    def get_all_app_users():
```

File: homepage/models.py [hunk 5]
```diff
@@ -109,11 +110,39 @@ def does_course_have_prerequisites(course):
...
+        return list(AppUser.objects.all())
+
+    @staticmethod
+    def find_appUser(username_ToFind):
```
Comment (kobihk) [inline homepage/models.py]: maybe 
```suggestion
    def get_app_user(username):

```

File: homepage/tests/appUser_model_tests.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import pytest
+from homepage.models import AppUser
+
+
+@pytest.mark.django_db
+class TestAppUser:
+
+    @pytest.fixture
+    def appUsers(self):
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]:
```suggestion
    def app_users(self):

```

File: homepage/tests/appUser_model_tests.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import pytest
+from homepage.models import AppUser
+
+
+@pytest.mark.django_db
+class TestAppUser:
+
+    @pytest.fixture
+    def appUsers(self):
+        appUser1 = AppUser.create_appUser('user1', 'user1@mta.ac.il', 'useruser222')
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: please fix all the occurrences according to:
https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables
```suggestion
        app_user1 = AppUser.create_appUser('user1', 'user1@mta.ac.il', 'useruser222')

```

File: homepage/tests/appUser_model_tests.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+
+@pytest.mark.django_db
+class TestAppUser:
+
+    @pytest.fixture
+    def appUsers(self):
+        appUser1 = AppUser.create_appUser('user1', 'user1@mta.ac.il', 'useruser222')
+        appUser2 = AppUser.create_appUser('user2', 'user2@mta.ac.il', '1234')
+        appUser3 = AppUser.create_appUser('user3', 'user3@gmail.com', '10101010')
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: same ^^

File: homepage/tests/appUser_model_tests.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+        return (appUser1, appUser2, appUser3)
+
+    def test_create_appUser(self, appUsers):
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]:
```suggestion
    def test_create_app_user(self, app_users):

```

File: homepage/tests/appUser_model_tests.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+        # 1. The 3 appUsers which are created in the fixture above are the correct type of object (AppUser)
+        # 2. The 3 appUsers which are created in the fixture above were sucesfully saved in the DB
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]:
```suggestion
        # The 3 appUsers created in the fixture above are:
        # 1. The correct type of object (AppUser)
        # 2. Successfully saved in the DB

```

File: homepage/tests/appUser_model_tests.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            assert isinstance(appUser, AppUser)
+            assert AppUser.objects.filter(pk=appUser.pk).exists()
+
+        # 3. An appUser with invalid username is not created
+        try:
+            AppUser.create_appUser('', 'user4@mta.ac.il', 'useruser222')
+        except ValueError:
+            invalid_username_created = False
+
+        assert invalid_username_created is False
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: please separate this method into few tests 
each one test separate scenario:
1. test_valid_app_users
2. test_invald_username
3. test_invalid_password

File: homepage/tests/appUser_model_tests.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        assert (appUser_invalid_password.user).has_usable_password() is False
+
+    def test_toggle_user_activation(self, appUsers):
+        assert appUsers[0].user.is_active is True
+
+        appUsers[0].toggle_user_activation()
+        assert appUsers[0].user.is_active is False
+
+        appUsers[0].toggle_user_activation()
+        assert appUsers[0].user.is_active is True
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: I would split it into 2 tests:
1. test_is_active_user
2. test_toggle_user_activation

usually, we will try to do one assert per one test 
and toggle is a good example of 2 asserts in 1 test

File: homepage/tests/appUser_model_tests.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+        for appUser in appUsers:
+            assert appUser in all_AppUsers
+
+    def test_find_appUser(self, appUsers):
+        assert (AppUser.find_appUser('user1')
+                and AppUser.find_appUser('user2')
+                and AppUser.find_appUser('user3'))
+
+        assert AppUser.find_appUser('unknown_user') is None
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: please split it into 2 tests:
1. test_get_app_user (you can do it with only one like 'user2')
2. test_get_undefined_app_user

**Review by kobihk — CHANGES_REQUESTED** [review]

File: homepage/models.py (hunk 3 above)
Comment (LeonRabinovich) [reply homepage/models.py]: > 
> 
> maybe you can create new AppUser object by:

I thought about this, but decided against it for 2 reasons :

1. I don't think the constructor of the class should save the object in the DB, this is why there is a separate method with a unique name (create_app_user) which indicates that an app_user **is created and also saved in the DB**

2. Django built-in User class also implements this in a similar way (and not with _init_)-
`User.objects.create_user()` : creates the user and saves it in the DB 
I think its more readable to implement it similarly to Django

File: homepage/models.py (hunk 4 above)
Comment (LeonRabinovich) [reply homepage/models.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 6 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 9 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 10 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 12 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/models.py (hunk 5 above)
Comment (LeonRabinovich) [reply homepage/models.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 7 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: fixed

File: homepage/tests/appUser_model_tests.py (hunk 11 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 13 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/models.py [hunk 14]
```diff
@@ -109,11 +110,40 @@ def does_course_have_prerequisites(course):
...
+    def toggle_user_activation(self):
+        self.user.is_active = not(self.user.is_active)
+
+    @staticmethod
+    # Creates a new app_user (without superuser permmisions) and saves it in the DB
+    # The is_active attribute of AppUser.user is automaticly set to 'True' when creating a new AppUser
+    def create_app_user(username, email, password):
+        app_user = AppUser()
+        user = User.objects.create_user(username, email, password)
+        app_user.user = user
```
Comment (kobihk) [inline homepage/models.py]:
```suggestion
        app_user.user = User.objects.create_user(username, email, password)

```

File: homepage/models.py [hunk 15]
```diff
@@ -109,11 +110,40 @@ def does_course_have_prerequisites(course):
...
+
+        app_user.save()
+        return app_user
+
+    @staticmethod
+    def get_all_app_users():
+        return list(AppUser.objects.all())
+
+    @staticmethod
+    def get_app_user(username_ToFind):
```
Comment (kobihk) [inline homepage/models.py]: if you have a method like here get_app_user and it gets username as a parameter so IMHO you don't need to add the suffix "_to_find"
```suggestion
    def get_app_user(username):

```

File: homepage/tests/appUser_model_tests.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        assert invalid_username_created is False
+
+    # An app_user with empty password does not have a usuable password (Django built in create_user behaviour)
+    def test_invalid_password(self, app_users):
+        app_user_invalid_password = AppUser.create_app_user('user6', 'user6@mta.ac.il', None)
+        assert (app_user_invalid_password.user).has_usable_password() is False
+
+    def test_is_active_user(self, app_users):
+        for app_user in app_users:
+            assert app_user.user.is_active is True
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]:
```suggestion
            assert app_user.user.is_active

```

File: homepage/tests/appUser_model_tests.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+            assert AppUser.objects.filter(pk=app_user.pk).exists()
+
+    # An app_user with invalid username is not created
+    def test_invald_username(self, app_users):
+        try:
+            AppUser.create_app_user('', 'user4@mta.ac.il', 'useruser222')
+        except ValueError:
+            invalid_username_created = False
+
+        assert invalid_username_created is False
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py:29]: you can do it like:
```suggestion
        assert not invalid_username_created

```
but if it's better for you it is also ok :)

**Review by kobihk — CHANGES_REQUESTED** [review]

File: homepage/models.py (hunk 14 above)
Comment (LeonRabinovich) [reply homepage/models.py]: changed

File: homepage/models.py (hunk 15 above)
Comment (LeonRabinovich) [reply homepage/models.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 16 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: changed

File: homepage/tests/appUser_model_tests.py (hunk 17 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py:29]: i like it better with False

File: homepage/models.py [hunk 18]
```diff
@@ -109,11 +110,39 @@ def does_course_have_prerequisites(course):
...
+        self.user.is_active = not(self.user.is_active)
+
+    @staticmethod
+    # Creates a new app_user (without superuser permmisions) and saves it in the DB
+    # The is_active attribute of AppUser.user is automaticly set to 'True' when creating a new AppUser
+    def create_app_user(username, email, password):
+        app_user = AppUser()
+        app_user.user = User.objects.create_user(username, email, password)
+        app_user.save()
+
```
Comment (kobihk) [inline homepage/models.py]: IMHO you can remove the blank line before the last line of the method

File: homepage/models.py [hunk 19]
```diff
@@ -109,11 +110,39 @@ def does_course_have_prerequisites(course):
...
+    @staticmethod
+    def get_all_app_users():
+        return list(AppUser.objects.all())
+
+    @staticmethod
+    def get_app_user(username):
+        try:
+            user = User.objects.get(username=username)
+            return AppUser.objects.get(user=user)
+
```
Comment (kobihk) [inline homepage/models.py]: same ^^

File: homepage/tests/appUser_model_tests.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+@pytest.mark.django_db
+class TestAppUser:
+
+    @pytest.fixture
+    def app_users(self):
+        app_user1 = AppUser.create_app_user('user1', 'user1@mta.ac.il', 'useruser222')
+        app_user2 = AppUser.create_app_user('user2', 'user2@mta.ac.il', '1234')
+        app_user3 = AppUser.create_app_user('user3', 'user3@gmail.com', '10101010')
+
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: same ^^

File: homepage/tests/appUser_model_tests.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+            assert isinstance(app_user, AppUser)
+            assert AppUser.objects.filter(pk=app_user.pk).exists()
+
+    # An app_user with invalid username is not created
+    def test_invald_username(self, app_users):
+        try:
+            AppUser.create_app_user('', 'user4@mta.ac.il', 'useruser222')
+        except ValueError:
+            invalid_username_created = False
+
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: same ^^

File: homepage/tests/appUser_model_tests.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        for app_user in app_users:
+            app_user.toggle_user_activation()
+            assert app_user.user.is_active is False
+
+            app_user.toggle_user_activation()
+            assert app_user.user.is_active
+
+    def test_get_all_app_users(self, app_users):
+        all_AppUsers = AppUser.get_all_app_users()
+
```
Comment (kobihk) [inline homepage/tests/appUser_model_tests.py]: same ^^

**Review by kobihk — CHANGES_REQUESTED** [review]

File: homepage/models.py (hunk 18 above)
Comment (LeonRabinovich) [reply homepage/models.py]: Removed

File: homepage/models.py (hunk 19 above)
Comment (LeonRabinovich) [reply homepage/models.py]: Removed

File: homepage/tests/appUser_model_tests.py (hunk 20 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: Removed

File: homepage/tests/appUser_model_tests.py (hunk 21 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: Removed

File: homepage/tests/appUser_model_tests.py (hunk 22 above)
Comment (LeonRabinovich) [reply homepage/tests/appUser_model_tests.py]: Removed

**Review by kobihk — APPROVED** [review]

**Pull request merged**